keps/sig-scheduling/20200116-coscheduling-plugin-based-on-scheduler-framework.md

```diff
@@ -0,0 +1,216 @@
+type Coscheduling struct {
+    FrameworkHandle     	framework.FrameworkHandle
+    PodLister               corelisters.PodLister
+    // Key is the name of PodGroup.
+    PodGroupInfos       	map[string]PodGroupInfo
+    // Name of the last scheduled podgroup
+    LastPodGroup        	string
+}
+ 
+type PodGroupInfo struct {
+    // LastFailureTimestamp stores the timestamp of last scheduling failure.
+    LastFailureTimestamp 	time.Time
+    UID           			types.UID
+    MinAvailable  		int
+    Name                                 String
+}
+```
+
+1.  `PodGroupInfo` is initialized the first time the pod belongs to the PodGroup is encountered, and LastFailureTimestamp is updated every time the PodGroup fails to schedule.
+2.  `LastPodGroup` records which `PodGroup` the last scheduled pod belongs to.
+3.  `UID` is the unique identification value used to distinguish different podgroups.
+ 
+
+### Extension points
+
+#### QueueSort
+In order to make the pods which belongs to the same `PodGroup` to be scheduled together as much as possible, implement a strategy in `QueueSort` phase. 
+
+```go
+  func  Less(podA *PodInfo, podB *PodInfo) bool
+```
+1. When priorities (i.e. .spec.priorityValue) are different, they are compared based on their priorities. When priorities are the same, they are operated according to the following process.
+   
+2. When podA and podB are both regularPods (we will check it by their labels), it follows the same logic of default in-tree [PrioritySort](https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/framework/plugins/queuesort/priority_sort.go#L41-L45) plugin.
+   
+3. When podA is regularPod, podB is pgPod. they are compared according to podA’s timestamp and podB’s `LastFailureTimestamp` (we get the PodGroupInfo from the cache). If timestampA is earlier than LastFailureTimestampB, return true. Otherwise, return false. When podA is pgPod, podB is regularPod. Ditto
+   
+4. When podA and podB are both pgPods. Compare the the `LastFailureTimestamp` of the podGroup. 
+   1. If `LastFailureTimestampA` is earlier than `LastFailureTimestampB`, return true. `LastFailureTimestampA` is later than `LastFailureTimestampB`, return false.
+   
+   2. If `LastFailureTimestampA` is equal to `LastFailureTimestampB`, we will compare the `UID` of `PodGroup`. (The purpose is to distinguish different `PodGroup` with the same `LastFailureTimestamp` and to keep the pods of the same `PodGroup` together)
+
+**Note1**: There are different `LastFailureTimestamp` (even if they are the same, the UID will be different). So when the pods enter the queue, the pods that belongs to the same PodGroup will be together.
+
+#### Pre-Filter
+1. When a pod which belongs to a `PodGroup` comes, calculate the total number of Pods  excludes terminating ones belongs to the same PodGroup. If the result is less than minAvailable, the scheduling process is terminated directly.
```

Comment:
```suggestion
1. When a pod which belongs to a `PodGroup` comes, calculate the total number of Pods that belong to the same PodGroup excluding terminating ones. If the result is less than minAvailable, the scheduling process is terminated directly.
```

Reply:
as in PreFilter returns unschedulable? note that this will cause a scheduling failure not a "fit error": see here: https://github.com/kubernetes/kubernetes/blob/e6b5194ec12a7aa241c702019d6cd6f360b232d5/pkg/scheduler/core/generic_scheduler.go#L187

Reply:
I am not sure I understand what you are tying to achieve, so similar to the comments above, start by describing the goal that PreFilter is trying to achieve not its implementation. For example: "PreFilter validates that ...."

Reply:
`PreFilter` validates that if the total number of pods belonging to the same `PodGroup` is less than `minAvailable`. If so, the scheduling process will be interrupted directly. 
I change it to return `UnschedulableAndUnresolvable` in PreFilter here.

Reply:
It doesn't matter whether the return status is ```UnschedulableAndUnresolvable``` or ```Unschedulable```, in both cases the scheduler will return a normal error, not a fit error, look at the code I pointed to in the comment above.

Reply:
em.. @ahg-g  I think it is not a fit error, maybe a normal error.  For example, the current total number of all pods belongs to a same `PodGroup` is 3 (Maybe there are other reasons why the left pods haven't been created yet), but the `minAvailable` of the `PodGroup` is 5. I think it is better to abort the scheduling cycle  in `PreFilter `.

Reply:
I think it's meant to fail quickly for the first N-1 unassigned Pods, while recording the number in the cache. Then when the Nth Pod comes, scheduler knows there are N-1 Pods pending in the internal queue, so it proceeds the Nth Pod with the reguarlar scheduling cycle. Is it right? @denkensk